### PR TITLE
Documentation site using mkdocs

### PR DESCRIPTION
**Pull request by TomPham97**
This will probably take at least a week if that's ok for you. I'll most likely ask lots of questions since I'm pretty much learning mkdocs from scratch 😅

### CODE REVIEW

**Comment by JoaoLages** [comment]
> This will probably take at least a week if that's ok for you. I'll most likely ask lots of questions since I'm pretty much learning mkdocs from scratch 😅

Yeah, no hurries! Actually I'll be learning with/from you because I never used mkdocs 😂

File: README.md [hunk 1]
```diff
@@ -1,11 +1,11 @@
-<div align="center">
+<!-- <div align="center"> -->
 
 # Diffusers-Interpret 🤗🧨🕵️‍♀️
 
 ![PyPI Latest Package Version](https://img.shields.io/pypi/v/diffusers-interpret?logo=pypi&style=flat&color=orange) ![GitHub License](https://img.shields.io/github/license/JoaoLages/diffusers-interpret?logo=github&style=flat&color=green) 
 
 `diffusers-interpret` is a model explainability tool built on top of [🤗 Diffusers](https://github.com/huggingface/diffusers)
-</div>
+<!-- </div> -->
```
Comment (TomPham97) [inline README.md]: I'm not sure why, but the center alignment syntax isn't displayed properly in Mkdocs. I'm temporarily disabling this and will look for a solution later.

File: mkdocs.yml [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    - navigation.top
+    - navigation.tabs
+  palette:
+  # Palette toggle for automatic mode
+  - media: "(prefers-color-scheme)"
+    primary: yellow
+    accent: red
+    toggle:
+      icon: material/brightness-auto
+      name: Switch to light mode
```
Comment (TomPham97) [inline mkdocs.yml]: The [automatic theme mode](https://squidfunk.github.io/mkdocs-material/setup/changing-the-colors/#automatic-light-dark-mode) appears to be sponsors-only. Maybe I should comment it out.

**Comment by TomPham97** [comment]
So the first part of the site should be ready to publish. Apparently you just have to merge and run `mkdocs gh-deploy`, according to [this](https://www.mkdocs.org/user-guide/deploying-your-docs/). I'm still deciding on the way to present the API and am quite a big fan of [this layout](https://pandas.pydata.org/docs/reference/frame.html). I did some experiments here and there but wasn't completely satisfied with my decision. Let me know what you think in general!

**Comment by TomPham97** [comment]
I based the majority of customizations on [squidfunk](https://squidfunk.github.io/mkdocs-material/). Another helpful resource is the [PyMdown Extension](https://facelessuser.github.io/pymdown-extensions/extensions/arithmatex/).

**Comment by TomPham97** [comment]
> So the first part of the site should be ready to publish. Apparently you just have to merge and run `mkdocs gh-deploy`, according to [this](https://www.mkdocs.org/user-guide/deploying-your-docs/). I'm still deciding on the way to present the API and am quite a big fan of [this layout](https://pandas.pydata.org/docs/reference/frame.html). I did some experiments here and there but wasn't completely satisfied with my decision. Let me know what you think in general!

So I ran the `mkdocs gh-deploy` command already, thus you won't need to do so after merging this PR ☺️. [This](https://tompham97.github.io/diffusers-interpret/) is what the page looks like from my fork.

File: README.md (hunk 1 above)
Comment (JoaoLages) [reply README.md]: Yes, we should fix this before merging the PR 👍

**Comment by JoaoLages** [comment]
> > So the first part of the site should be ready to publish. Apparently you just have to merge and run `mkdocs gh-deploy`, according to [this](https://www.mkdocs.org/user-guide/deploying-your-docs/). I'm still deciding on the way to present the API and am quite a big fan of [this layout](https://pandas.pydata.org/docs/reference/frame.html). I did some experiments here and there but wasn't completely satisfied with my decision. Let me know what you think in general!
> 
> So I ran the `mkdocs gh-deploy` command already, thus you won't need to do so after merging this PR ☺️. [This](https://tompham97.github.io/diffusers-interpret/) is what the page looks like from my fork.

The page looks really cool! I really like the layout you chose! ⭐ 

I have a couple of suggestions:

- Use [readthedocs](https://readthedocs.org/) to host the website
- Instead of reading from README, let's create a `docs/source`  directory:
  - Example: [diffusers' main page](https://github.com/huggingface/diffusers/blob/main/docs/source/index.mdx) - this is overkill, we have a simpler package :)
  - Then they also have `autodoc` to build parts of the documentation from docstrings in code. [Check this example](https://github.com/huggingface/diffusers/blob/main/docs/source/api/pipelines/stable_diffusion.mdx#stablediffusionpipelineoutput). I actually don't know if `autodoc` is something from [their own doc-builder](https://github.com/huggingface/doc-builder) or from `mkdocs`, but there should be a way to have documentation from docstrings.
 
You can start with a dummy main page that grabs docstring from the code and I will help you filling it 😁

File: README.md (hunk 1 above)
Comment (JoaoLages) [reply README.md]: If we read from `docs/source`, this will not be an issue

**Review by TomPham97 — COMMENTED** [review]
[At the moment, only Sphinx supports additional formats. `pdf`, `epub`, and `htmlzip` outputs are not yet supported when using MkDocs.](https://docs.readthedocs.io/en/stable/config-file/v2.html#:~:text=At%20the%20moment%2C%20only%20Sphinx%20supports%20additional%20formats.%20pdf%2C%20epub%2C%20and%20htmlzip%20output%20is%20not%20yet%20supported%20when%20using%20MkDocs.)

I'm just leaving this configuration set to `all` for future-proofing. The build process on readthedocs appears to be uneffected.

**Comment by TomPham97** [comment]
> I have a couple of suggestions:
> 
> * Use [readthedocs](https://readthedocs.org/) to host the website

I have set up the configuration files necessary for **readthedocs**. It would be the most convenient if you import the project yourself by [connecting your readthedocs account to Github](https://docs.readthedocs.io/en/stable/connected-accounts.html) or [registering](https://readthedocs.org/accounts/signup/) using your Github account.

To include me as a maintainer of the **readthedocs** project, please add my username `TomPham97` into the *Add maintainer* field. This is located in the [*Maintainers* submenu](https://readthedocs.org/dashboard/diffusers-interpret/users/) of the *Admin* tab.

**Comment by TomPham97** [comment]
> * I actually don't know if `autodoc` is something from [their own doc-builder](https://github.com/huggingface/doc-builder) or from `mkdocs`, but there should be a way to have documentation from docstrings.

Autodocs appears to [require HF's doc-builder](https://github.com/huggingface/transformers/tree/main/docs#writing-documentation---specification:~:text=Once%20you%20have%20setup%20the%20doc%2Dbuilder%20and%20additional%20packages%2C%20you%20can%20generate%20the%20documentation%20by%20typing%20the%20following%20command%3A). However, I'll double check once more tomorrow to make sure if it could work as a standalone tool. [Mkdocstrings](https://mkdocstrings.github.io/usage/#) is already available with the current setup and can be called with a tripple colon `:::` followed by the python script/function name. Maybe we could start with the `Txt2Img explainer` and see whether mkdocstrings is sufficient? What do you think?

File: docs/api/explainer.md [hunk 3]
```diff
@@ -0,0 +1 @@
+::: explainer
```
Comment (TomPham97) [inline docs/api/explainer.md:1]: [Here](https://tompham97.github.io/diffusers-interpret/api/explainer/)'s an example of what mkdocstrings has pulled from `explainer.py`.

File: mkdocs.yml [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  - search
+  - git-revision-date-localized:
+      enable_creation_date: true
+
+watch:
+  - src/diffusers_interpret
+
+nav:
+  - index.md
+  - explainer: api/explainer.md
```
Comment (TomPham97) [inline mkdocs.yml:65]: The site layout is configured here.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+jobs:
+  deploy:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-python@v2
+        with:
+          python-version: 3.x
+      - run: pip install mkdocs "mkdocstrings[python]" mkdocs-material 
+      - run: mkdocs gh-deploy --force
```
Comment (TomPham97) [inline .github/workflows/ci.yml]: Just in case the PR merge has some issues. In order for this workflow to be implemented, your [personal access token](https://github.com/settings/tokens) needs to authenticate workflows as well.

**Comment by TomPham97** [comment]
> * Instead of reading from README, let's create a `docs/source`  directory

I'll start a new PR for this, so the current PR doesn't get too cluttered.

**Comment by JoaoLages** [comment]
> `TomPham97`

Done! This is very easy when you send all the needed links! 😂 🙏

**Comment by JoaoLages** [comment]
>I have set up the configuration files necessary for readthedocs. It would be the most convenient if you import the project yourself by [connecting your readthedocs account to Github](https://docs.readthedocs.io/en/stable/connected-accounts.html) or [registering](https://readthedocs.org/accounts/signup/) using your Github account.

>To include me as a maintainer of the readthedocs project, please add my username TomPham97 into the Add maintainer field. This is located in the [Maintainers submenu](https://readthedocs.org/dashboard/diffusers-interpret/users/) of the Admin tab.

 Done! This is very easy when you send all the needed links! 😂 🙏

File: docs/api/explainer.md (hunk 3 above)
Comment (JoaoLages) [reply docs/api/explainer.md:1]: not a bad start! I like the documentation style. We can definitely adapt the code formatting and docstrings for them to look better in the documentation.
Do you know why other methods were not imported to the documentation? For instance, [the __init__ method](https://github.com/JoaoLages/diffusers-interpret/blob/2dd01d6a494dd0bc18b5c2fa99dda7132ae03f42/src/diffusers_interpret/explainer.py#L27), [gradient checkpointing](https://github.com/JoaoLages/diffusers-interpret/blob/2dd01d6a494dd0bc18b5c2fa99dda7132ae03f42/src/diffusers_interpret/explainer.py#L333) or [_mimic_pipeline_call](https://github.com/JoaoLages/diffusers-interpret/blob/2dd01d6a494dd0bc18b5c2fa99dda7132ae03f42/src/diffusers_interpret/explainer.py#L349). `_mimic_pipeline_call` is actually a private method, so idk if we should add it to the docs, need to check what are the common practices

**Comment by JoaoLages** [comment]
> > * I actually don't know if `autodoc` is something from [their own doc-builder](https://github.com/huggingface/doc-builder) or from `mkdocs`, but there should be a way to have documentation from docstrings.
> 
> Autodocs appears to [require HF's doc-builder](https://github.com/huggingface/transformers/tree/main/docs#writing-documentation---specification:~:text=Once%20you%20have%20setup%20the%20doc%2Dbuilder%20and%20additional%20packages%2C%20you%20can%20generate%20the%20documentation%20by%20typing%20the%20following%20command%3A). However, I'll double check once more tomorrow to make sure if it could work as a standalone tool. [Mkdocstrings](https://mkdocstrings.github.io/usage/#) is already available with the current setup and can be called with a tripple colon `:::` followed by the python script/function name. Maybe we could start with the `Txt2Img explainer` and see whether mkdocstrings is sufficient? What do you think?

Definitely! From [this example](https://tompham97.github.io/diffusers-interpret/api/explainer/#explainer.BasePipelineImg2ImgExplainer) you sent, I think we can definitely work only with mkdocstrings
Thanks a lot for all your effort, amazing work! ⭐ 🚀

File: docs/api/explainer.md (hunk 3 above)
Comment (TomPham97) [reply docs/api/explainer.md:1]: I tried to specify the functions by calling `::: explainer.BasePipelineExplainer.__init__`, but it didn't work as intended. I'm looking into the issue but will need another day or two before I could update you with the findings ☺️

Comment (TomPham97) [reply docs/api/explainer.md:1]: > `_mimic_pipeline_call` is actually a private method, so idk if we should add it to the docs, need to check what are the common practices

I'll actually need to study public and private methods, as well as related terminologies. I'm still learning Python as a beginner and would appreciate any of your recommendations for resource 😅

Comment (JoaoLages) [reply docs/api/explainer.md:1]: private methods don't really exist in Python, it's just a convention from other Object-oriented programming languages as Java
Something private is something that should not be accessed outside the class object.
For example: `explainer._mimic_pipeline_call` should not be called, as it is a private method only used inside the class object. 
Once again, just a convention because in Python you access everything!

Comment (TomPham97) [reply docs/api/explainer.md:1]: > Do you know why other methods were not imported to the documentation? For instance, [the **init** method](https://github.com/JoaoLages/diffusers-interpret/blob/2dd01d6a494dd0bc18b5c2fa99dda7132ae03f42/src/diffusers_interpret/explainer.py#L27), [gradient checkpointing](https://github.com/JoaoLages/diffusers-interpret/blob/2dd01d6a494dd0bc18b5c2fa99dda7132ae03f42/src/diffusers_interpret/explainer.py#L333) or [_mimic_pipeline_call](https://github.com/JoaoLages/diffusers-interpret/blob/2dd01d6a494dd0bc18b5c2fa99dda7132ae03f42/src/diffusers_interpret/explainer.py#L349).

So a method needs to have an appropriately-formatted docstring in order to be imported into the documentation. Here are a few things that I've discovered thus far:
- The docstring cannot be just empty pairs of quotations `''' '''`.
- Any sentences in the beginning of a docstring will be formatted as [the description](https://tompham97.github.io/diffusers-interpret/api/explainer/#explainer.BasePipelineImg2ImgExplainer:~:text=%C2%B6-,Calls%20a%20DiffusionPipeline%20and%20generates%20explanations%20for%20a%20given%20prompt.,-Parameters%3A).
- Arguments included under `Args: ` will show up in a [table of parameters](https://tompham97.github.io/diffusers-interpret/api/explainer/#explainer.BasePipelineImg2ImgExplainer:~:text=Parameters%3A,%7B%7D).
- Likewise, things under `Returns: ` will show up in the [returns table](https://tompham97.github.io/diffusers-interpret/api/explainer/#explainer.BasePipelineImg2ImgExplainer:~:text=%7B%7D-,Returns%3A,%5D%20if%20init_image%20is%20not%20None%20and%20explanation_2d_bounding_box%20is%20not%20None,-Source%20code%20in)
- Private methods with a single underscore in the front will not be included.
- Methods are listed alphabetically, with characters such as `_` being listed first. For instance, `__call__` would be before `__init__`, which is before `def_...`.
- If there's a specific `@` that belongs to a method, it would also be included in the documentation.

So long story short, you just need to include docstrings in the correct format in order for methods like `__init__` to appear. What do you think about my explanation? Does it make sense to you?

File: .github/workflows/ci.yml (hunk 5 above)
Comment (JoaoLages) [reply .github/workflows/ci.yml]: I missed this comment! What should I do here?

File: docs/api/explainer.md (hunk 3 above)
Comment (JoaoLages) [reply docs/api/explainer.md:1]: Yup, totally makes sense! 👍

**Comment by JoaoLages** [comment]
Looking at [https://tompham97.github.io/diffusers-interpret/api/explainer/#explainer.BasePipelineExplainer.__call__](https://tompham97.github.io/diffusers-interpret/api/explainer/#explainer.BasePipelineExplainer.__call__), it seems that:

1. We need to add a docstring to `__init__` because it does not appear here - also, is it possible for `__init__` to appear before `__call__`? it would make sense but not super important
image

2. Can we format this part
image
To appear like this?
image
It would make it easier to read. No big issue if not possible.

3. Seems like we need to remove ` from the types. Also, is it possible to put some links here? For example a link for the `AttributionMethods` class. Also, we can remove the `*optional*` part since we have the `Default` column
image

4. Return type is bugged, how do we fix it?
image

**Comment by JoaoLages** [comment]
Also, this scrollbar makes the documentation hard to read. Can we change the template to get rid of the horizontal scrollbar?

image

File: .github/workflows/ci.yml (hunk 5 above)
Comment (TomPham97) [reply .github/workflows/ci.yml]: I'm incorporating a github workflow into this pull request so the mkdocs site gets built automatically whenever there's a change in the repository. However, that means your personal access token needs to be able to authenticate workflows. I wrote the original message to let you know about this just in case the merge encounters an error.

**Comment by TomPham97** [comment]
I'm currently on vacation and will attempt to answer your questions in 3 days or so. Thank you for your patience ☺️

File: .github/workflows/ci.yml (hunk 5 above)
Comment (JoaoLages) [reply .github/workflows/ci.yml]: Created the access token for GitHub workflows, do I need to set it somewhere?

Comment (TomPham97) [reply .github/workflows/ci.yml]: I just modified my existing access token to allow for GitHub workflows. I suggest you try to merge this PR first and see if the workflow here is allowed.

**Comment by TomPham97** [comment]
> Looking at [https://tompham97.github.io/diffusers-interpret/api/explainer/#explainer.BasePipelineExplainer.**call**](https://tompham97.github.io/diffusers-interpret/api/explainer/#explainer.BasePipelineExplainer.__call__), it seems that:
> 
> 1. We need to add a docstring to `__init__` because it does not appear here - also, is it possible for `__init__` to appear before `__call__`? it would make sense but not super important
>    image
> 2. Can we format this part
>    image
>    To appear like this?
>    image
>    It would make it easier to read. No big issue if not possible.
> 3. Seems like we need to remove `from the types. Also, is it possible to put some links here? For example a link for the`AttributionMethods`class. Also, we can remove the`_optional_`part since we have the`Default` column
>    image
> 4. Return type is bugged, how do we fix it?
>    image

So I discovered the helpful [options](https://mkdocstrings.github.io/python/usage/#globallocal-options) for rendering the docstring. I'm currently doing some tests to verify their functionalities.

**Comment by TomPham97** [comment]
> 1. We need to add a docstring to `__init__` because it does not appear here - also, is it possible for `__init__` to appear before `__call__`? it would make sense but not super important

There's indeed [this option](https://mkdocstrings.github.io/python/usage/#:~:text=members_order,%22alphabetical%22.) that you're looking for. To display the `__init__` method, the docstring from `explainer.py` needs to be formatted appropriately, which is addressed in PR #20 . For now, it would be best if you could merge this PR as a base for the documentation site.

**Comment by TomPham97** [comment]
> Also, this scrollbar makes the documentation hard to read. Can we change the template to get rid of the horizontal scrollbar?
> 
> image

I've set [this template option](https://mkdocstrings.github.io/python/usage/#:~:text=docstring_section_style,%22table%22.) to `list` instead of `table`. You can view the change [here](https://tompham97.github.io/diffusers-interpret/api/explainer/).
**If you don't see any changes, I suggest opening the link in an igcognito window to reset the browser's previous cache.**

**Comment by TomPham97** [comment]
> 2\. Can we format this part
>     image
>     To appear like this?
>     image
>     It would make it easier to read. No big issue if not possible.

Is [this option](https://mkdocstrings.github.io/python/usage/#:~:text=show_signature_annotations,Default%3A%20False.) what you were looking for? Setting it to `true` instead of the default `false` will yield something like this:
Screen Shot 2022-10-05 at 16 23 26

**Comment by JoaoLages** [comment]
> > Also, this scrollbar makes the documentation hard to read. Can we change the template to get rid of the horizontal scrollbar?
> > image
> 
> I've set [this template option](https://mkdocstrings.github.io/python/usage/#:~:text=docstring_section_style,%22table%22.) to `list` instead of `table`. You can view the change [here](https://tompham97.github.io/diffusers-interpret/api/explainer/). **If you don't see any changes, I suggest opening the link in an igcognito window to reset the browser's previous cache.**

Nice! I prefer it this way!

**Comment by JoaoLages** [comment]
> > 2. Can we format this part
> > image
> > To appear like this?
> > image
> > It would make it easier to read. No big issue if not possible.
> 
> Is [this option](https://mkdocstrings.github.io/python/usage/#:~:text=show_signature_annotations,Default%3A%20False.) what you were looking for? Setting it to `true` instead of the default `false` will yield something like this: Screen Shot 2022-10-05 at 16 23 26

better! can we add a `\n` for each parameter?

**Review by JoaoLages — APPROVED** [review]
I'm approving this as you requested for further developments :)
Great job!

Do whatever changes you need and merge it
